collect_app/src/main/java/org/odk/collect/android/http/CollectServerClient.java

```diff
@@ -0,0 +1,157 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.odk.collect.android.http;
+
+import android.net.Uri;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.kxml2.io.KXmlParser;
+import org.kxml2.kdom.Document;
+import org.odk.collect.android.utilities.DocumentFetchResult;
+import org.odk.collect.android.utilities.WebCredentialsUtils;
+import org.xmlpull.v1.XmlPullParser;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.HashMap;
+
+import javax.inject.Inject;
+
+import timber.log.Timber;
+
+/**
+ * Common utility methods for managing the credentials associated with the
+ * request context and constructing http context, client and request with the
+ * proper parameters and OpenRosa headers.
+ *
+ * @author mitchellsundt@gmail.com
+ */
+public class CollectServerClient {
+
+    private static final String HTTP_CONTENT_TYPE_TEXT_XML = "text/xml";
+
+    protected OpenRosaHttpInterface httpInterface;
+
+    @Inject
+    public CollectServerClient(OpenRosaHttpInterface httpInterface) {
+        this.httpInterface = httpInterface;
+    }
+
+
+    /**
+     * Common method for returning a parsed xml document given a url and the
+     * http context and client objects involved in the web connection.
+     */
+    public DocumentFetchResult getXmlDocument(String urlString) {
+
+        // parse response
+        Document doc;
+
+        HttpGetResult inputStreamResult;
+        try {
+            inputStreamResult = getHttpInputStream(urlString, HTTP_CONTENT_TYPE_TEXT_XML);
+
+            InputStream is = inputStreamResult.getInputStream();
+            InputStreamReader isr = null;
+            try {
+                isr = new InputStreamReader(is, "UTF-8");
+                doc = new Document();
+                KXmlParser parser = new KXmlParser();
+                parser.setInput(isr);
+                parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true);
+                doc.parse(parser);
+                isr.close();
+                isr = null;
+            } finally {
+                if (isr != null) {
+                    try {
+                        // ensure stream is consumed...
+                        final long count = 1024L;
+                        while (isr.skip(count) == count) {
+                            // skipping to the end of the http entity
+                        }
+
+                        isr.close();
```

Comment:
We can switch out these try blocks with try with resources statements!

Reply:
Great suggestion!